sdk/core/Azure.Core/src/DynamicData/DynamicData.Converters.cs

```diff
@@ -142,5 +143,19 @@ private class UnixTimeDateTimeOffsetConverter : JsonConverter<DateTimeOffset>
                 writer.WriteNumberValue(value);
             }
         }
+
+        private class BinaryDataConverter : JsonConverter<BinaryData>
+        {
+            public override BinaryData Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                byte[] value = reader.GetBytesFromBase64();
```

Comment:
99% of the time this is probably correct, but some services use base64url or - even more rarely - hex. What could we do in those cases? Should we add something to our `DynamicDataOptions` (or whatever it is - already has a couple properties, IIRC) to vary this? Though, I'm not even sure that would solve the problem since it can vary from model property to property within a service. Key Vault, for example, *mostly* uses base64url but does use base64 in a couple of places.

I won't hold up the PR for this alone, but maybe open a tracking issue. Perhaps we don't tackle it until it becomes a problem, but I suspect that might not take long. base64url is uncommon, but not exactly rare.

Reply:
@heaths, could you share an example of the KeyVault cases?

There has been some discussion of adding a ContentType property to BinaryData, and if that were to manifest, we could handle all of this under the hood.  That would have the positive side-effect of not requiring customers to pass another enum value, where they could potentially make a mistake with it.

Would this address the `base64url` case as well?

Reply:
See repo:Azure/azure-sdk-for-net path:sdk/keyvault/Azure base64url.

base64url is not a content type - more of an encoding analogous to `Content-Encoding` if you're thinking of HTTP headers. Swagger supports different representations of byte arrays, and any converter would need to support different encodings or DD will not work for those services' models.

Honestly, without the swagger I'm not sure how you'd even know but perhaps,

1. Throw a `NotSupportedException` that wraps whatever the converter throws so it's consistent, and document what that could mean.
2. Support some sort of type->converter mapping like @m-nash has been talking about for model de/serialziation. Of course, that may be overkill and it's better just to use models at that point.
3. Allow the SDKs to plug in some "hooks" to help DD. They know best.

To get some sense of scale, see https://github.com/search?q=repo%3AAzure%2Fazure-rest-api-specs%20path%3Aspecification%20base64url&type=code vs. a search for "base64" (though the latter can be encoded in swagger/typespecs differently, and typically mentioned in descriptions as well). Seems it's limited to Key Vault and Attestation currently.